### PR TITLE
[Annotation plugin] Make getting annotation manager configurable.

### PR DESCRIPTION
**Pull request by Chaoba**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [x] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [ ] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog>Make getting annotation manager configurable</changelog>`.

### Summary of changes

In this pr:

- Add a configure class for annotation plugin, users can custom layer id and source id, etc. while getting the annotation manager instance.
- Fix the issue that annotation will be erased while changing style 
- Update test demo to very changes
QQ20210125-205122

### CODE REVIEW

File: app/src/main/java/com/mapbox/maps/testapp/examples/annotation/CircleActivity.kt [hunk 1]
```diff
@@ -103,4 +97,24 @@ class CircleActivity : AppCompatActivity() {
     super.onDestroy()
     mapView.onDestroy()
   }
+
+  companion object {
+    /** Current index of style*/
+    private var index: Int = 0
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/annotation/CircleActivity.kt:103]: Why is it in companion object? Can't it be regular field?

Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/annotation/CircleActivity.kt:103]: Applies to all similar code below.

File: app/src/main/java/com/mapbox/maps/testapp/examples/annotation/SymbolActivity.kt [hunk 2]
```diff
@@ -40,26 +39,25 @@ class SymbolActivity : AppCompatActivity() {
   private var symbolManager: SymbolManager? = null
   private var symbol: Symbol? = null
   private val animators: MutableList<ValueAnimator> = mutableListOf()
-
+  private var airPortImage: Bitmap? = null
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/annotation/SymbolActivity.kt:42]:
```suggestion
  private var airportImage: Bitmap? = null
```

File: app/src/main/java/com/mapbox/maps/testapp/examples/annotation/SymbolActivity.kt [hunk 3]
```diff
@@ -89,7 +87,7 @@ class SymbolActivity : AppCompatActivity() {
 
         // create nearby symbols
         val nearbyOptions: SymbolOptions = SymbolOptions()
-          .withPoint(Point.fromLngLat(0.367099, 6.626384))
+          .withPoint(Point.fromLngLat(0.367099, 6.526384))
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/annotation/SymbolActivity.kt:90]: Values to const

File: app/src/main/java/com/mapbox/maps/testapp/examples/annotation/Utils.kt [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import java.io.IOException
+import java.io.InputStreamReader
+import java.io.Reader
+import java.nio.charset.Charset
+import java.util.*
+
+/**
+ * Useful utilities used throughout the testapp.
+ */
+object Utils {
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/annotation/Utils.kt:17]: Perhaps `AnnotationUtils`? For better search in project.

File: app/src/main/java/com/mapbox/maps/testapp/examples/annotation/Utils.kt [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+   */
+  fun createRandomPoint(): Point {
+    val random = Random()
+    return Point.fromLngLat(
+      random.nextDouble() * -360.0 + 180.0,
+      random.nextDouble() * -180.0 + 90.0
+    )
+  }
+
+  @Throws(IOException::class)
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/annotation/Utils.kt:79]: Better wrap code inside with `try catch` here than throw exception upstream and catch it in each usage case.

File: app/src/main/java/com/mapbox/maps/testapp/examples/annotation/Utils.kt [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  fun loadStringFromAssets(context: Context, fileName: String): String {
+    if (TextUtils.isEmpty(fileName)) {
+      throw NullPointerException("No GeoJSON File Name passed in.")
+    }
+    val `is` = context.assets.open(fileName)
+    val rd = BufferedReader(InputStreamReader(`is`, Charset.forName("UTF-8")))
+    return readAll(rd)
+  }
+
+  @Throws(IOException::class)
```
Comment (kiryldz) [inline app/src/main/java/com/mapbox/maps/testapp/examples/annotation/Utils.kt:89]: Better wrap code inside with try catch here than throw exception upstream and catch it in each usage case.

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt [hunk 7]
```diff
@@ -105,29 +117,42 @@ abstract class AnnotationManagerImpl<G : Geometry, T : Annotation<G>, S : Annota
...
 
   protected fun initLayerAndSource() {
-    initializeDataDrivenPropertyMap()
-    source = createSource()
-    layer = createLayer()
-    if (belowLayerId == null) {
-      style.addLayer(layer)
-    } else {
-      style.addLayerBelow(layer, belowLayerId)
+    if (layer == null || source == null) {
```
Comment (kiryldz) [inline plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:135]: Just to confirm - we recreate `source` even it was __not__ `null` but `layer == null`. Is it expected behavior?

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt [hunk 8]
```diff
@@ -428,21 +459,23 @@ abstract class AnnotationManagerImpl<G : Geometry, T : Annotation<G>, S : Annota
...
-          callback.onQueryAnnotation(annotations[id])
+    if (layer != null) {
+      mapFeatureQueryDelegate.queryRenderedFeatures(
+        screenCoordinate,
+        RenderedQueryOptions(
+          listOf(layer!!.layerId),
+          literal(true)
+        )
+      ) { features ->
+        features.value?.let { featureList ->
```
Comment (kiryldz) [inline plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:470]: If `feature.value == null` - we will not call `callback.onQueryAnnotation(null)`. Is it expected behavior?

File: sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationConfig.kt [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+/**
+ * Config class for annotation plugin to get the annotation manager
+ */
+data class AnnotationConfig(
+  /** The id of the layer above the annotation layer*/
+  val belowLayerId: String? = null,
+  /** The layer id for layer inside of annotation manager, if not set will use the default id managed by annotation manager.*/
+  val layerId: String? = null,
+  /** The source id for layer inside of annotation manager, if not set will use the default id managed by annotation manager.*/
+  val sourceId: String? = null,
```
Comment (kiryldz) [inline sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationConfig.kt:12]:
```suggestion
  val sourceId: String? = null
```

File: sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationConfig.kt [hunk 10]
```diff
@@ -0,0 +1,13 @@
+package com.mapbox.maps.plugin.annotation
+
+/**
+ * Config class for annotation plugin to get the annotation manager
+ */
+data class AnnotationConfig(
+  /** The id of the layer above the annotation layer*/
+  val belowLayerId: String? = null,
```
Comment (kiryldz) [inline sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationConfig.kt:8]: The doc states that `id of the layer above the annotation layer` while val is named `belowLayerId` 🤔

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt (hunk 7 above)
Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:135]: layer and source are created at the same time, here check them all just a double confirm.

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt (hunk 8 above)
Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:470]: Yes, everywhere get the result will check null. Have thought about it, it seems we don't need the callback interface to return a nullable result.

File: sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationConfig.kt (hunk 10 above)
Comment (Chaoba) [reply sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationConfig.kt:8]: We will add the annotation layer below the given layer, which is the same as the definition in Layer

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt [hunk 11]
```diff
@@ -186,14 +211,20 @@ abstract class AnnotationManagerImpl<G : Geometry, T : Annotation<G>, S : Annota
...
-      annotation
+    if (!styleStateDelegate.isFullyLoaded()) {
+      return
     }
-    source.featureCollection(FeatureCollection.fromFeatures(features))
-    if (style.getSource(source.sourceId) == null) {
-      style.addSource(source)
+    source?.let {
+      if (!style.styleSourceExists(it.sourceId)) {
+        return
```
Comment (pengdev) [inline plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:219]: shall we add some warning logs?

Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:219]: Sounds good.

**Review by kiryldz — APPROVED** [review]
@Chaoba - great job, LGTM!
Also please add changelog entry to PR description.

**Pull request merged**